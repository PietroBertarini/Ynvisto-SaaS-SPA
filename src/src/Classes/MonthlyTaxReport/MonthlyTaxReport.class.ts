import {
  ICommonOpsReport,
  IDayTradeReport, IDefaultReport,
  IExemptVolumeInfo,
  IFiiReport, IMonthlyReportsApiModel,
  IProfitInfo,
  ITaxInfo,
} from './MonthlyTaxReport.types';
import { undefinedToZero, negativeOrUndefinedToZero } from '../../Util/general.utils';

export default class MonthlyTaxReport {
  /**
   * All info about exempt value sale of brazilian stocks.
   */
  readonly exemptVolumeInfo: IExemptVolumeInfo;

  /**
   * Tax value of each report.
   */
  readonly taxInfo: ITaxInfo;

  /**
   * Profit value of each report.
   */
  readonly profitInfo: IProfitInfo;

  /**
   *  Marks if a DARF is paid or not
   */
  readonly darfPaymentId: number | null;

  /**
   * The month of this report (0 indexed. January = 0, December = 11)
   */
  readonly monthIndex: number;

  /**
   * The Year of this report
   */
  readonly year: number;

  /**
   * Taxes owed due to stock market operations (Common operations and Day Trade operations), net from IRRF
   */
  readonly stockMarketNetTaxes: number;

  /**
   * DARF Taxes for the current month
   */
  readonly darfCurrentMonthTaxes: number;

  /**
   * Amount that should be paid this month via DARF, respecting the minimum payable
   * rules and carry-over rules of the Receita Federal
   */
  readonly darfPayableTaxes: number;

  /**
   *  Report for common operations
   */
  readonly commonOpsReport?: ICommonOpsReport;

  /**
   *  Report for day trade
   */
  readonly dayTradeReport?: IDayTradeReport;

  /**
   * Report for Crypto
   */
  readonly fiiReport?: IFiiReport;

  /**
   * Net taxes of all the assets taxed via the DARF mechanism
   */
  readonly accumulatedDarfTaxes: number;

  /**
   * Report for Crypto
   */
  readonly cryptoReport?: IDefaultReport;

  /**
   * Report for Subscription rights
   */
  readonly subscriptionRightReport?: IDefaultReport;

  /**
   * Report for international assets
   */
  readonly internationalReport?: IDefaultReport;

  constructor(
    exemptVolumeInfo : IExemptVolumeInfo,
    taxInfo: ITaxInfo,
    profitInfo: IProfitInfo,
    month: number,
    year: number,
    stockMarketNetTaxes: number,
    darfPaymentId: number | null,
    accumulatedDarfTaxes: number,
    darfPayableTaxes: number,
    darfCurrentMonthTaxes: number,
    commonOpsReport?: ICommonOpsReport,
    dayTradeReport?: IDayTradeReport,
    fiiReport?: IFiiReport,
    cryptoReport?: IDefaultReport,
    subscriptionRightReport?: IDefaultReport,
    internationalReport?: IDefaultReport,
  ) {
    this.exemptVolumeInfo = exemptVolumeInfo;
    this.taxInfo = taxInfo;
    this.profitInfo = profitInfo;
    this.darfPaymentId = darfPaymentId;
    this.monthIndex = month;
    this.year = year;
    this.stockMarketNetTaxes = stockMarketNetTaxes;
    this.commonOpsReport = commonOpsReport;
    this.dayTradeReport = dayTradeReport;
    this.fiiReport = fiiReport;
    this.cryptoReport = cryptoReport;
    this.subscriptionRightReport = subscriptionRightReport;
    this.internationalReport = internationalReport;
    this.accumulatedDarfTaxes = accumulatedDarfTaxes;
    this.darfPayableTaxes = darfPayableTaxes;
    this.darfCurrentMonthTaxes = darfCurrentMonthTaxes;
  }

  /**
   * Construct the Monthly Tax Report object that is consumed at SPA(Single Page App).
   * @param apiModel
   */
  static fromApiModel(apiModel: IMonthlyReportsApiModel) {
    const date = new Date(Date.parse(apiModel.date));
    const exemptVolumeInfo = getExemptVolumeInfo(apiModel?.commonOpsReport);
    const taxInfo = getTaxInfo(apiModel);
    const profitInfo = getProfitInfo(apiModel);
    const month = date.getMonth();
    const year = date.getFullYear();
    const internationalReport = undefined;

    const {
      commonOpsReport,
      stockMarketNetTaxes,
      dayTradeReport,
      fiiReport,
      cryptoReport,
      subscriptionRightReport,
      accumulatedDarfTaxes,
      darfPayableTaxes,
      darfCurrentMonthTaxes,
    } = apiModel;

    return new MonthlyTaxReport(
      exemptVolumeInfo,
      taxInfo,
      profitInfo,
      month,
      year,
      stockMarketNetTaxes,
      apiModel.darfPaymentId,
      accumulatedDarfTaxes,
      darfPayableTaxes,
      darfCurrentMonthTaxes,
      commonOpsReport,
      dayTradeReport,
      fiiReport,
      cryptoReport,
      subscriptionRightReport,
      internationalReport,
    );
  }

  /**
   * The sum of all profits.
   */
  get profit() {
    return (
      this.profitInfo.commonOpsProfit
      + this.profitInfo.dayTradeProfit
      + this.profitInfo.fiiProfit
      + this.profitInfo.cryptoProfit
      + this.profitInfo.internationalProfit
      + this.profitInfo.subscriptionRightProfit
    );
  }

  /**
   * The sum of previous' DARFs with less than R$10,00 not paid yet.
   */
  get previousDarf() {
    return (
      this.accumulatedDarfTaxes
      - this.darfCurrentMonthTaxes
    );
  }

  /**
   * The sum of all taxes.
   */
  get tax() {
    return (
      this.taxInfo.commonOpsTax
      + this.taxInfo.dayTradeTax
      + this.taxInfo.fiiTax
      + this.taxInfo.cryptoTax
      + this.taxInfo.internationalTax
      + this.taxInfo.subscriptionRightTax
      + this.previousDarf
    );
  }

  get remainExemptionVolumeBrazil() {
    return (negativeOrUndefinedToZero(this.exemptVolumeInfo.stockExemptionSaleThresholdBrazil
      - this.exemptVolumeInfo.stocksExemptionSaleVolumeUsedBrazil)
    );
  }
}

/**
 * Get the information about exempt volume Sale commons ops in brazilian stocks.
 */
function getExemptVolumeInfo(commonOpsReport?: ICommonOpsReport): IExemptVolumeInfo {
  return {
    stocksExemptionSaleVolumeUsedBrazil: negativeOrUndefinedToZero(commonOpsReport?.stocksExemptionSaleVolumeUsed),
    stockExemptionSaleThresholdBrazil: negativeOrUndefinedToZero(commonOpsReport?.stockExemptionSaleThreshold),
  };
}

/**
 * Get the final profit for each report from api.
 */
function getProfitInfo(apiModel: IMonthlyReportsApiModel): IProfitInfo {
  return {
    commonOpsProfit: undefinedToZero(apiModel?.commonOpsReport?.profits),
    dayTradeProfit: undefinedToZero(apiModel?.dayTradeReport?.profits),
    fiiProfit: undefinedToZero(apiModel?.fiiReport?.profits),
    cryptoProfit: undefinedToZero(apiModel?.cryptoReport?.profits),
    subscriptionRightProfit: undefinedToZero(apiModel?.subscriptionRightReport?.profits),
    internationalProfit: undefinedToZero(apiModel?.internationalReport?.profits),
  };
}
export interface StockMarketTaxes {
    /**
     * Taxes, net from all deductions, owed due to common ops.
     */
    commonOpsNetTaxes : number;
    /**
     * Taxes, net from all deductions, owed due to day trading
     */
    dayTradeNetTaxes : number;

}

/**
 * Get the proportional tax of dayTrade and common ops report.(Based on stockMarketTaxes )
 */
export function getStockMarketTaxes(
  commonOpsReport?: ICommonOpsReport, dayTradeReport ?: IDayTradeReport, stockMarketNetTaxes?: number,
) : StockMarketTaxes {
  const zeroStockMarketTaxes = {
    commonOpsNetTaxes: 0,
    dayTradeNetTaxes: 0,
  };

  const commonTaxOwed = undefinedToZero(commonOpsReport?.taxOwed);
  const dayTradeTaxOwed = undefinedToZero(dayTradeReport?.taxesOwed);
  const stockMarketTaxes = undefinedToZero(stockMarketNetTaxes);
  const totalTaxOwed = commonTaxOwed + dayTradeTaxOwed;
  if (totalTaxOwed === 0) {
    return zeroStockMarketTaxes;
  }
  return {
    commonOpsNetTaxes: stockMarketTaxes * (commonTaxOwed / totalTaxOwed),
    dayTradeNetTaxes: stockMarketTaxes * (dayTradeTaxOwed / totalTaxOwed),
  };
}

/**
 * Get the final tax for each report from api.
 */
function getTaxInfo(apiModel: IMonthlyReportsApiModel): ITaxInfo {
  const { commonOpsNetTaxes, dayTradeNetTaxes } = getStockMarketTaxes(apiModel.commonOpsReport,
    apiModel.dayTradeReport,
    apiModel.stockMarketNetTaxes);
  return {
    commonOpsTax: commonOpsNetTaxes,
    dayTradeTax: dayTradeNetTaxes,
    fiiTax: undefinedToZero(apiModel?.fiiReport?.netTaxes),
    cryptoTax: undefinedToZero(apiModel?.cryptoReport?.taxesOwed),
    subscriptionRightTax: undefinedToZero(apiModel?.subscriptionRightReport?.taxesOwed),
    internationalTax: undefinedToZero(apiModel?.internationalReport?.taxesOwed),
  };
}
