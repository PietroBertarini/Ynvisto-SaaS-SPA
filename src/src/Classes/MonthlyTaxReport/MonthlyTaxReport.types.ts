export interface IExemptVolumeInfo {
    stocksExemptionSaleVolumeUsedBrazil: number;
    stockExemptionSaleThresholdBrazil: number;
}
export interface ITaxInfo {
    commonOpsTax: number;
    dayTradeTax: number;
    fiiTax: number;
    cryptoTax: number;
    subscriptionRightTax: number;
    internationalTax: number;
}

export interface IProfitInfo {
    commonOpsProfit: number;
    dayTradeProfit: number;
    fiiProfit: number;
    cryptoProfit: number;
    subscriptionRightProfit: number;
    internationalProfit: number;
}

export interface IDayTradeReport {
    taxesOwed: number,
    profits: number,
    previousMonthLosses: number,
    irrf: number,
    irrfToCompensate: number,
    taxableProfits: number,
    taxBase: number,
}
export interface IFiiReport {
    profits: number,
    previousMonthLosses: number,
    netTaxes: number,
    taxBase: number,
}
export interface IDefaultReport {
    profits: number,
    saleVolume: number,
    taxesOwed: number,
}

export interface ICommonOpsReport {
    stocksExemptionSaleVolumeUsed: number,
    stockExemptionSaleThreshold: number,
    totalSaleVolume: number,
    taxOwed: number,
    profits: number,
    previousMonthLosses: number,
    irrf: number,
    irrfToCompensate: number,
    taxableProfits: number,
    taxBase: number,
}
export interface IMonthlyReportsApiModel {
  /**
   * .NET Date string. i.e: "2020-11-03T00:00:00"
   */
    date: string,
    darfPaymentId: number | null,
    commonOpsReport?: ICommonOpsReportApiModel,
    dayTradeReport?: IDayTradeReportApiModel,
    stockMarketNetTaxes: number,
    fiiReport?: IFiiReportApiModel,
    cryptoReport?: IDefaultReportApiModel,
    subscriptionRightReport?: IDefaultReportApiModel,
    internationalReport?: IDefaultReportApiModel,
    accumulatedDarfTaxes: number,
    darfPayableTaxes: number,
    darfCurrentMonthTaxes: number,
}
export interface IFiiReportApiModel {
    profits: number,
    previousMonthLosses: number,
    netTaxes: number,
    taxBase: number,
}

export interface ICommonOpsReportApiModel {
    stocksExemptionSaleVolumeUsed: number,
    stockExemptionSaleThreshold: number,
    totalSaleVolume: number,
    taxOwed: number,
    profits: number,
    previousMonthLosses: number,
    irrf: number,
    irrfToCompensate: number,
    taxableProfits: number,
    taxBase: number,
}

export interface IDayTradeReportApiModel {
    taxesOwed: number,
    profits: number,
    previousMonthLosses: number,
    irrf: number,
    irrfToCompensate: number,
    taxableProfits: number,
    taxBase: number,
}
export interface IDefaultReportApiModel {
    profits: number,
    saleVolume: number,
    taxesOwed: number,
}
