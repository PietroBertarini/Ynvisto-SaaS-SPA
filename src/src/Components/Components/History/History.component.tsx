import React from 'react';
import { format } from 'date-fns';
import ColorfulCircle from '../ColorfulCircle/ColorfulCircle.component';
import YnvistoCard from '../YnvistoCard/YnvistoCard.component';
import { extractPercentageToCircle, HistoryProp, getType } from './History.utils';
import {
  Main, TimeLine, TimeBlock, InnerCard,
} from './History.styles';

const History = ({ data }: HistoryProp) => (
  <Main>
    {data && data.map((timeBlock) => (
      <TimeBlock key={timeBlock.date}>
        <div className="dateLine">
          <ColorfulCircle
            firstColor="#05E985"
            lastColor="#EB5050"
            percentage={extractPercentageToCircle(timeBlock.operations)}
          />
          <p>{format(new Date(timeBlock.date), 'dd/MM/yyyy')}</p>
        </div>
        <TimeLine>
          <div className="cards">
            {timeBlock.operations.map((operation) => (
              <YnvistoCard key={operation.volume} color={operation.profit < 0 ? '#8C2F2F' : '#029B58'}>
                <InnerCard color={operation.profit < 0 ? '#EB5050' : '#05E985'}>
                  <h2>
                    {operation.operationType === 'Purchase' ? 'Compra' : 'Venda'}
                    {' '}
                    de
                    {' '}
                    {operation.ticker}
                    {' '}
                    :
                    {' '}
                    {getType(operation.assetType, operation.isDayTrade)}
                  </h2>
                  <div className="infoCard">
                    <div className="infoLarge">
                      <h5>Total:</h5>
                      <p>{operation.volume.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}</p>
                    </div>
                    <div className="info">
                      <h5>Preço:</h5>
                      <p>{operation.price.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}</p>
                    </div>
                    <div className="info">
                      <h5>Unidades:</h5>
                      <p>{operation.quantity}</p>
                    </div>
                    <div className="info">
                      <h5>{operation.isDayTrade ? 'Preço de Compra:' : 'Preço Médio:'}</h5>
                      <p>{operation.avgCost.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}</p>
                    </div>
                    <div className="infoLarge">
                      <h5>Lucro:</h5>
                      <p>{operation.profit.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}</p>
                    </div>
                  </div>
                </InnerCard>
              </YnvistoCard>
            ))}
          </div>
        </TimeLine>
      </TimeBlock>
    ))}
  </Main>
);

export default History;
