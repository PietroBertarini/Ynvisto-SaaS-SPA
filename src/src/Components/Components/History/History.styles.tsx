import styled from 'styled-components';

interface CardColorProp {
  color: string;
}

export const Main = styled.div`
  margin: 5vh;
  display: flex;
  flex-direction: column;
  flex: 1;
  align-items: center;

  h3 {
    color: #fff;
    font-weight: bold;
    font-size: 5vw;
    margin-bottom: 20px;
    text-align: center;
    margin-left: 239px;
  }
`;

export const TimeLine = styled.div`
  min-height: 10vh;
  min-width: 45vw;
  margin-left: 15px;
  border-left: 3px solid #B1B1B1;
  padding-left: 1vw;

  .cards {
    padding: 2vh;
    max-width: 32vw;
  }
`;

export const TimeBlock = styled.div`
  .dateLine {
    height: 35px;
    display: flex;
    
    p {
      color: #B1B1B1;
      font-family: 'Zurich';
      font-weight: bold;
      height: 35px;
      padding: 0;
      margin: 0 0 0 30px;
      font-size: 30px;
    }
  }
`;

export const InnerCard = styled.div<CardColorProp>`
  padding: 0 5px;

  h2 {
    padding: 8px 0 0 2.5vw;
    color: #fff;
    font-weight: bold;
    font-size: 0.8vw;
    margin-bottom: 20px;
  }

  .infoCard {
    padding: 0 0 10px 10px;
    display: flex;
    flex-direction: row;
    flex: 1;

    .info {
      margin-right: 20px;
      min-width: 3vw;

      h5 {
        color: rgba(255, 255, 255, 0.64);
        font-size: 0.7vw;
        padding: 0;
        margin: 0;
      }

      p {
        font-size: 0.9vw;
        color: ${(props) => props.color};
        padding: 0;
        margin: 0;
      }
    }

    .infoLarge {
      margin-right: 20px;
      min-width: 5vw;

      h5 {
        color: rgba(255, 255, 255, 0.64);
        font-size: 0.7vw;
        padding: 0;
        margin: 0;
      }

      p {
        font-size: 0.9vw;
        color: ${(props) => props.color};
        padding: 0;
        margin: 0;
      }
    }
  }
`;
