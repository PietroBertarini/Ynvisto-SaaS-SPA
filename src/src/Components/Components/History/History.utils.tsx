import { IParsedOperationsHistory } from '../../Pages/MonthlyTaxReport/MonthlyTaxReport.utils';
import { IOperationsHistory } from '../../../Redux/Services/TaxReports/TaxReports.actions';

export interface HistoryProp {
  data: IParsedOperationsHistory[],
}

/**
 * get the porcentage of profits of total.
 * @param operations: operations of specified day.
 */
export function extractPercentageToCircle(operations: IOperationsHistory[]): number {
  let losses = 0;
  let profits = 0;

  operations.forEach((operation) => {
    if (operation.profit >= 0) profits += operation.profit;
    else losses -= operation.profit;
  });

  if (profits === 0 && losses === 0) return 100;
  return (profits * 100) / (profits + losses);
}

export function getType(type: string, isDayTrade: boolean): string {
  const types = {
    Acoes: 'Ações',
    Etf: 'ETF',
    Options: 'Opções',
    Crypto: 'Criptomoeda',
    Fii: 'FII',
    BDR: 'BDR',
    SubscriptionRight: 'Direito de Subscrição',
    International: 'Ação Internacional',
  };

  if (types[type] === 'Ações' && isDayTrade) return 'Daytrade';
  if (types[type] === 'Ações') return 'Operação Comum';

  return types[type];
}
