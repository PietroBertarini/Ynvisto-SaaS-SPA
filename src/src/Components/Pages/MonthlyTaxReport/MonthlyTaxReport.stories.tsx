import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import MonthlyTaxReport from './MonthlyTaxReport.component';

export default {
  title: 'Services/MonthlyTaxReport',
  component: MonthlyTaxReport,
} as Meta;

const Template: Story<{}> = (args) => (<MonthlyTaxReport {...args} />
);

export const Default = Template.bind({});
