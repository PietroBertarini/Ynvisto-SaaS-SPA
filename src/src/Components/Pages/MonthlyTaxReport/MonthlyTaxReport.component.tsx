import React, { useCallback, useEffect, useState } from 'react';
import { Grid, Typography } from '@material-ui/core';
import { connect, ConnectedProps, useDispatch } from 'react-redux';
import '../../../Theme/sweetalert2/sweetalert2.scss';
import { useHistory, useParams } from 'react-router';
import { Redirect } from 'react-router-dom';
import { State } from '../../../Redux/root-reducer';
import TaxDetailsModal from '../../Components/TaxDetailsModal/TaxDetailsModal.component';
import ScrollTabs from '../../Components/ScrollTabs/ScrollTabs.component';
import {
  fetchTaxReportsRequest,
} from '../../../Redux/Services/TaxReports/TaxReports.actions';
import Spinner from '../../Components/Spinner/spinner.component';
import {
  getExemptionSaleVolumeUsedInfo, getExemptionSaleVolumeUsedTitleInfo,
  getMonthNumber, getMonthlyReport,
  getProfitData,
  getProfitHeaderValue,
  getExemptionTitleStatus,
  getTaxData,
  getTaxHeaderValue, getProfitTitleStatus, getTaxesTitleStatus,
  handleClickDarf, deleteDarf, markDarfAsPaid, parseOperationsHistory,
} from './MonthlyTaxReport.utils';
import MonthlyCardReport from '../../Components/MonthlyCardReport/MonthlyCardReport.component';
import { allTabs } from '../TaxReport/TaxReport.utils';
import { getTaxDetailsInfo } from '../../Components/MonthlyCardReport/MonthlyCardReport.utils';
import YnvistoButton, { EButtonColor } from '../../Components/YnvistoButton/YnvistoButton.component';
import History from '../../Components/History/History.component';
import { useStyles } from './MonthlyTaxReport.styles';

/**
 * Get the stores states that TaxComponents uses and observe.
 * @param state: the stored stated of the Ynvisto.
 */
const mapStateToProps = (state: State) => ({
  taxReport: state.taxReports.taxReport,
  isFetching: state.taxReports.isFetchingReport,
  spreadsheetAdm: state.authorization.spreadsheetAdmin,
  error: state.taxReports.error,
  user: state.authorization.user,
  irpf: state.irpf.assetsRights,
});

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>
type Props = PropsFromRedux

function MonthlyTaxReport(props: Props) {
  const dispatch = useDispatch();
  const {
    taxReport, isFetching, error, spreadsheetAdm, user, irpf,
  } = props;
  console.log(irpf);
  const classes = useStyles();

  const { month: navBarMonth } = useParams < { month: string }>();
  const history = useHistory();
  const navBarMonthNumber = getMonthNumber(navBarMonth);

  const monthlyTaxReport = getMonthlyReport(navBarMonthNumber, taxReport);
  const currentDate = new Date();

  const [markingDarf, setMarkingDarf] = useState(false);

  const historyData = parseOperationsHistory(taxReport?.operations, navBarMonthNumber);

  useEffect(() => {
    if (taxReport === undefined) {
      dispatch(fetchTaxReportsRequest(spreadsheetAdm));
    }

    return () => {
      // Clean previously data for fetch again yhe most updated tax report data.
      // dispatch(taxReportCleanData());
    };
  }, [dispatch, taxReport, spreadsheetAdm]);

  const monthIndex = monthlyTaxReport?.monthIndex ?? -1;
  const year = monthlyTaxReport?.year ?? -1;
  const darfId = monthlyTaxReport?.darfPaymentId ?? 0;

  const payDarf = useCallback(async () => {
    if (markingDarf) {
      return; // Noop while the DARF request is loading
    }

    if (!monthlyTaxReport) {
      console.warn('Cannot confirm darf payment for an empty report');
      return;
    }

    // Starts the loader
    setMarkingDarf(true);

    await markDarfAsPaid(monthIndex, year, dispatch);

    setMarkingDarf(false);
  }, [markingDarf, monthlyTaxReport, monthIndex, year, dispatch]);

  const deleteDarfCallback = useCallback(async () => {
    if (markingDarf) {
      return; // Noop while the DARF request is loading
    }

    if (darfId === 0) {
      console.warn('Cannot delete nonexistent darf payment');
      return;
    }

    // Start the loader
    setMarkingDarf(true);

    await deleteDarf(monthIndex, year, darfId, dispatch);

    setMarkingDarf(false);
  }, [darfId, dispatch, markingDarf, monthIndex, year]);

  if (isFetching) {
    return <Spinner />;
  }

  // if user put a month that doesn't exist, go back to Tax Report Page
  if (navBarMonthNumber === 0) {
    history.push('/Lucros/');
  }

  // TODO: Find a way to make this check standard across all API Calls, just like the error building and checking is
  if (error?.httpStatusCode === 401) {
    return <Redirect to="/Logout/" />;
  }

  const monthlyTaxReportDetails = getTaxDetailsInfo(monthlyTaxReport);
  const exemptionSaleVolumeUsedData = getExemptionSaleVolumeUsedInfo(monthlyTaxReport);
  const taxData = getTaxData(monthlyTaxReport);
  const profitData = getProfitData(monthlyTaxReport);
  const exemptionSaleVolumeUsedHeader = getExemptionSaleVolumeUsedTitleInfo(monthlyTaxReport);
  const taxHeader = getTaxHeaderValue(monthlyTaxReport);
  const profitHeader = getProfitHeaderValue(monthlyTaxReport);

  const renderButtonsSection = () => {
    if (!monthlyTaxReport) {
      return null;
    }

    const darfPayable = monthlyTaxReport.darfPayableTaxes;
    const hasDarf = navBarMonthNumber < (currentDate.getMonth() + 1) && darfPayable > 0;
    if (!hasDarf) {
      return null;
    }

    const { darfPaymentId } = monthlyTaxReport;
    const darfButton = darfPaymentId
      ? (
        <YnvistoButton
          status={EButtonColor.DANGER}
          title="Remover pagamento de DARF"
          onClick={deleteDarfCallback}
          isLoading={markingDarf}
        />
      )
      : (
        <YnvistoButton
          title="Confirmar pagamento de DARF"
          onClick={payDarf}
          isLoading={markingDarf}
        />
      );

    return (
      <div className={classes.buttons}>
        {darfButton}
        <YnvistoButton
          title="Gerar DARF"
          onClick={() => handleClickDarf(navBarMonthNumber, darfPayable, user?.userName)}
        />
      </div>
    );
  };

  return (
    <div className={classes.root}>
      <Typography className={classes.header}>
        Apuração de Lucros
      </Typography>
      <ScrollTabs initialTabValue={navBarMonthNumber} tabsLabel={allTabs} />
      {renderButtonsSection()}
      <Grid container item xs={12} spacing={3} className={classes.grid}>
        <Grid item xs={4}>
          <MonthlyCardReport
            status={getExemptionTitleStatus(exemptionSaleVolumeUsedHeader)}
            title="Volume isento em operações comuns"
            headerMonetaryValue={exemptionSaleVolumeUsedHeader}
            data={exemptionSaleVolumeUsedData}
          />
        </Grid>
        <Grid item xs={4}>
          <MonthlyCardReport
            status={getTaxesTitleStatus(taxHeader)}
            title="Imposto a pagar"
            headerMonetaryValue={taxHeader}
            data={taxData}
            detailsModal={<TaxDetailsModal data={monthlyTaxReportDetails} />}
          />
        </Grid>
        <Grid item xs={4}>
          <MonthlyCardReport
            status={getProfitTitleStatus(profitHeader)}
            title={profitHeader >= 0 ? 'Lucro do mês' : 'Prejuízo do mês'}
            headerMonetaryValue={profitHeader}
            data={profitData}
          />
        </Grid>
      </Grid>
      <Typography variant="body1" className={classes.obs}>
        Obs: Impostos de Criptomoedas e Direito de subscriçaão não são
        incluidos na DARF segundo a Receita Federal, devem ser pagas no GCAP
      </Typography>
      <Typography className={classes.subheader}>
        Histórico de operações com lucro ou prejuízo em
        {' '}
        {navBarMonth}
      </Typography>
      {historyData.length !== 0
        ? <History data={historyData} />
        : (
          <Typography variant="body1" className={classes.message}>
            Você não possui operações com lucro ou prejuízo esse mês.
          </Typography>
        )}
    </div>
  );
}
export default connector(MonthlyTaxReport);
