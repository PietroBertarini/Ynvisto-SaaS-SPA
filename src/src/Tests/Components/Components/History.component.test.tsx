import React from 'react';
import { shallow } from 'enzyme';
import History from '../../../Components/Components/History/History.component';
import { ParsedOperationsHistoryNovember } from '../../../Mocks/MonthlyTaxReport.mock';

test('Renders History component and match to previous snapshot', () => {
  const component = shallow(<History data={ParsedOperationsHistoryNovember} />);
  expect(component).toMatchSnapshot();
});
