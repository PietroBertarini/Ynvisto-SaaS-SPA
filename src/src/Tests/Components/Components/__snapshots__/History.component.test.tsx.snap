// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders History component and match to previous snapshot 1`] = `
<styled.div>
  <styled.div
    key="2020-11-10T00:00:00"
  >
    <div
      className="dateLine"
    >
      <ColorfulCircle
        firstColor="#05E985"
        lastColor="#EB5050"
        percentage={100}
      />
      <p>
        10/11/2020
      </p>
    </div>
    <styled.div>
      <div
        className="cards"
      >
        <YnvistoCard
          color="#029B58"
          key="11432"
        >
          <styled.div
            color="#05E985"
          >
            <h2>
              Venda
               
              de
               
              HGLG11
               
              :
               
              FII
            </h2>
            <div
              className="infoCard"
            >
              <div
                className="infoLarge"
              >
                <h5>
                  Total:
                </h5>
                <p>
                  R$ 11.432,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço:
                </h5>
                <p>
                  R$ 114,32
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Unidades:
                </h5>
                <p>
                  100
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço Médio:
                </h5>
                <p>
                  R$ 113,32
                </p>
              </div>
              <div
                className="infoLarge"
              >
                <h5>
                  Lucro:
                </h5>
                <p>
                  R$ 100,00
                </p>
              </div>
            </div>
          </styled.div>
        </YnvistoCard>
      </div>
    </styled.div>
  </styled.div>
  <styled.div
    key="2020-11-11T00:00:00"
  >
    <div
      className="dateLine"
    >
      <ColorfulCircle
        firstColor="#05E985"
        lastColor="#EB5050"
        percentage={0}
      />
      <p>
        11/11/2020
      </p>
    </div>
    <styled.div>
      <div
        className="cards"
      >
        <YnvistoCard
          color="#8C2F2F"
          key="11432"
        >
          <styled.div
            color="#EB5050"
          >
            <h2>
              Venda
               
              de
               
              HGLG11
               
              :
               
              FII
            </h2>
            <div
              className="infoCard"
            >
              <div
                className="infoLarge"
              >
                <h5>
                  Total:
                </h5>
                <p>
                  R$ 11.432,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço:
                </h5>
                <p>
                  R$ 114,32
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Unidades:
                </h5>
                <p>
                  100
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço Médio:
                </h5>
                <p>
                  R$ 113,32
                </p>
              </div>
              <div
                className="infoLarge"
              >
                <h5>
                  Lucro:
                </h5>
                <p>
                  -R$ 100,00
                </p>
              </div>
            </div>
          </styled.div>
        </YnvistoCard>
      </div>
    </styled.div>
  </styled.div>
  <styled.div
    key="2020-11-12T00:00:00"
  >
    <div
      className="dateLine"
    >
      <ColorfulCircle
        firstColor="#05E985"
        lastColor="#EB5050"
        percentage={80}
      />
      <p>
        12/11/2020
      </p>
    </div>
    <styled.div>
      <div
        className="cards"
      >
        <YnvistoCard
          color="#029B58"
          key="500"
        >
          <styled.div
            color="#05E985"
          >
            <h2>
              Compra
               
              de
               
              OIBR3
               
              :
               
              Daytrade
            </h2>
            <div
              className="infoCard"
            >
              <div
                className="infoLarge"
              >
                <h5>
                  Total:
                </h5>
                <p>
                  R$ 500,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço:
                </h5>
                <p>
                  R$ 0,50
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Unidades:
                </h5>
                <p>
                  1000
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço de Compra:
                </h5>
                <p>
                  R$ 1,00
                </p>
              </div>
              <div
                className="infoLarge"
              >
                <h5>
                  Lucro:
                </h5>
                <p>
                  R$ 3.000,00
                </p>
              </div>
            </div>
          </styled.div>
        </YnvistoCard>
        <YnvistoCard
          color="#029B58"
          key="10000"
        >
          <styled.div
            color="#05E985"
          >
            <h2>
              Venda
               
              de
               
              BOVA11
               
              :
               
              ETF
            </h2>
            <div
              className="infoCard"
            >
              <div
                className="infoLarge"
              >
                <h5>
                  Total:
                </h5>
                <p>
                  R$ 10.000,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço:
                </h5>
                <p>
                  R$ 100,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Unidades:
                </h5>
                <p>
                  100
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço Médio:
                </h5>
                <p>
                  R$ 90,00
                </p>
              </div>
              <div
                className="infoLarge"
              >
                <h5>
                  Lucro:
                </h5>
                <p>
                  R$ 1.000,00
                </p>
              </div>
            </div>
          </styled.div>
        </YnvistoCard>
        <YnvistoCard
          color="#8C2F2F"
          key="500"
        >
          <styled.div
            color="#EB5050"
          >
            <h2>
              Compra
               
              de
               
              OIBR3
               
              :
               
              Daytrade
            </h2>
            <div
              className="infoCard"
            >
              <div
                className="infoLarge"
              >
                <h5>
                  Total:
                </h5>
                <p>
                  R$ 500,00
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço:
                </h5>
                <p>
                  R$ 0,50
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Unidades:
                </h5>
                <p>
                  1000
                </p>
              </div>
              <div
                className="info"
              >
                <h5>
                  Preço de Compra:
                </h5>
                <p>
                  R$ 1,00
                </p>
              </div>
              <div
                className="infoLarge"
              >
                <h5>
                  Lucro:
                </h5>
                <p>
                  -R$ 1.000,00
                </p>
              </div>
            </div>
          </styled.div>
        </YnvistoCard>
      </div>
    </styled.div>
  </styled.div>
</styled.div>
`;
